llvm/lib/Target/EraVM/EraVMInstrFormats.td

```diff
@@ -642,56 +712,102 @@ class IUMAr_r<EraVMOpcode opcode,
               dag outs, dag ins,
               string asmstr,
               list<dag> pattern>
- : IUMA<opcode, SrcReg, 0, outs, ins, asmstr, pattern> {
+ : IUMA<opcode, SrcSpecialReg, 0, outs, ins, asmstr, pattern> {
+  bits<4> rs0;
+  bits<4> rd0;
+
+  let Src0 = rs0;
+  let Dst0 = rd0;
 }
 
 class IUMAr_rr<EraVMOpcode opcode,
                dag outs, dag ins,
                string asmstr,
                list<dag> pattern>
- : IUMA<opcode, SrcReg, 1, outs, ins, asmstr, pattern> {
+ : IUMA<opcode, SrcSpecialReg, 1, outs, ins, asmstr, pattern> {
+  bits<4> rs0;
+  bits<4> rd0;
+  bits<4> rd1;
+
+  let Src0 = rs0;
+  let Dst0 = rd0;
+  let Dst1 = rd1;
 }
 
 class IUMAi_r<EraVMOpcode opcode,
               dag outs, dag ins,
               string asmstr,
               list<dag> pattern>
- : IUMA<opcode, SrcImm, 0, outs, ins, asmstr, pattern> {
+ : IUMA<opcode, SrcSpecialImm, 0, outs, ins, asmstr, pattern> {
+  bits<16> addr;
+  bits<4> rd0;
+
+  let Imm0 = addr;
+  let Dst0 = rd0;
 }
 
 class IUMAi_rr<EraVMOpcode opcode,
                dag outs, dag ins,
                string asmstr,
                list<dag> pattern>
- : IUMA<opcode, SrcImm, 1, outs, ins, asmstr, pattern> {
+ : IUMA<opcode, SrcSpecialImm, 1, outs, ins, asmstr, pattern> {
+  bits<16> addr;
+  bits<4> rd0;
+  bits<4> rd1;
+
+  let Imm0 = addr;
+  let Dst0 = rd0;
+  let Dst1 = rd1;
 }
 
 class IUMArr_<EraVMOpcode opcode,
               dag outs, dag ins,
               string asmstr,
               list<dag> pattern>
- : IUMA<opcode, SrcReg, 0, outs, ins, asmstr, pattern> {
+ : IUMA<opcode, SrcSpecialReg, 0, outs, ins, asmstr, pattern> {
+  bits<4> rs0;
+  bits<4> rs1;
+
+  let Src0 = rs0;
+  let Src1 = rs1;
 }
 
 class IUMArr_r<EraVMOpcode opcode,
                dag outs, dag ins,
                string asmstr,
                list<dag> pattern>
- : IUMA<opcode, SrcReg, 1, outs, ins, asmstr, pattern> {
+ : IUMA<opcode, SrcSpecialReg, 1, outs, ins, asmstr, pattern> {
+  bits<4> rs0;
+  bits<4> rs1;
+  bits<4> rd0;
+
+  let Src0 = rs0;
+  let Src1 = rs1;
+  let Dst0 = rd0;
 }
 
 class IUMAir_<EraVMOpcode opcode,
               dag outs, dag ins,
               string asmstr,
               list<dag> pattern>
- : IUMA<opcode, SrcImm, 0, outs, ins, asmstr, pattern> {
+ : IUMA<opcode, SrcSpecialImm, 0, outs, ins, asmstr, pattern> {
+  bits<16> addr;
+  bits<4> rs1;
+
+  let Imm0 = addr;
+  let Src1 = rs1;
 }
 
 class IUMAir_r<EraVMOpcode opcode,
                dag outs, dag ins,
                string asmstr,
                list<dag> pattern>
- : IUMA<opcode, SrcImm, 1, outs, ins, asmstr, pattern> {
-}
+ : IUMA<opcode, SrcSpecialImm, 1, outs, ins, asmstr, pattern> {
+  bits<16> addr;
+  bits<4> rs1;
+  bits<4> rd0;
 
-// TODO: There should be IUMAs / IUMAc
```

Comment:
If I understand correctly, IUMA class is only used for heap access instructions (i.e. pointer *dereference*, but not pointer *arithmetics*) that only accept `in_reg` / `in_regimm` and `out_reg` operands.

Reply:
Yes, it's correct.